File: skills/drawio/scripts/drawio.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
draw.io diagram tool -- create, read, and update .drawio.png files.

Creates PNG files with embedded draw.io XML (zTXt chunk), so the PNG
is both viewable as an image and editable in draw.io.

Requires `npx draw.io-export` on PATH for PNG rendering.

Usage:
    uv run drawio.py create --xml '<mxGraphModel>...</mxGraphModel>' -o diagram.drawio.png
    uv run drawio.py create --xml-file diagram.xml -o diagram.drawio.png
    uv run drawio.py read diagram.drawio.png
    uv run drawio.py update --xml '<mxGraphModel>...</mxGraphModel>' diagram.drawio.png
    uv run drawio.py update --xml-file updated.xml diagram.drawio.png
"""

from __future__ import annotations

import argparse
import json
import os
import struct
import subprocess
import sys
import tempfile
import urllib.parse
import zlib


# ============================================================
# PNG chunk helpers
#
# draw.io PNG format:
#   zTXt chunk key: "mxGraphModel"
#   zTXt value (after zlib inflate): <mxfile><diagram>PAYLOAD</diagram></mxfile>
#   PAYLOAD: base64( deflateRaw( urlEncode(mxGraphModelXml) ) )
#
# deflateRaw = raw deflate with NO zlib header (wbits=-15)
# ============================================================

import base64
import re


def _crc32(data: bytes) -> int:
    """Compute CRC32 for a PNG chunk (type + data)."""
    return zlib.crc32(data) & 0xFFFFFFFF


def _encode_diagram_payload(xml: str) -> str:
    """mxGraphModel XML -> URL-encode -> raw deflate -> base64."""
    url_encoded = urllib.parse.quote(xml, safe="")
    # wbits=-15 gives raw deflate (no zlib header)
    deflated = zlib.compress(url_encoded.encode("utf-8"), level=9, wbits=-15)
    return base64.b64encode(deflated).decode("ascii")


def _decode_diagram_payload(payload: str) -> str:
    """base64 -> raw inflate -> URL-decode -> mxGraphModel XML."""
    deflated = base64.b64decode(payload)
    # wbits=-15 for raw inflate
    url_encoded = zlib.decompress(deflated, wbits=-15).decode("utf-8")
    return urllib.parse.unquote(url_encoded)


def _build_ztxt_chunk(xml: str) -> bytes:
    """Build a PNG zTXt chunk embedding draw.io XML."""
    payload = _encode_diagram_payload(xml)
    mxfile_xml = f"<mxfile><diagram>{payload}</diagram></mxfile>"

    # zTXt format: key\0 compression_method(1 byte) compressed_data
    key = b"mxGraphModel\x00"
    compression_method = b"\x00"  # zlib
    compressed = zlib.compress(mxfile_xml.encode("utf-8"))
    chunk_data = key + compression_method + compressed

    chunk_type = b"zTXt"
    length = struct.pack(">I", len(chunk_data))
    crc = struct.pack(">I", _crc32(chunk_type + chunk_data))

    return length + chunk_type + chunk_data + crc


def embed_xml_in_png(png_path: str, xml: str) -> None:
    """Embed draw.io XML into a PNG file as a zTXt chunk."""
    data = _read_bytes(png_path)

    # Strip any existing mxGraphModel zTXt chunks first
    data = _strip_ztxt_chunks(data, key=b"mxGraphModel")

    chunk = _build_ztxt_chunk(xml)

    # Insert before IEND (last 12 bytes of a valid PNG)
    iend_start = len(data) - 12
    before = data[:iend_start]
    iend = data[iend_start:]

    with open(png_path, "wb") as f:
        f.write(before + chunk + iend)


def extract_xml_from_png(png_path: str) -> str | None:
    """Extract draw.io XML from a PNG file's zTXt metadata."""
    data = _read_bytes(png_path)

    # Walk PNG chunks looking for zTXt with key "mxGraphModel"
    pos = 8  # skip PNG signature
    while pos + 8 <= len(data):
        length = struct.unpack(">I", data[pos : pos + 4])[0]
        chunk_type = data[pos + 4 : pos + 8]
        chunk_data = data[pos + 8 : pos + 8 + length]

        if chunk_type == b"zTXt":
            null_pos = chunk_data.index(b"\x00")
            key = chunk_data[:null_pos]
            if key == b"mxGraphModel":
                # byte after null is compression method (0), then zlib data
                compressed = chunk_data[null_pos + 2 :]
                mxfile_xml = zlib.decompress(compressed).decode("utf-8")
                # Parse <mxfile><diagram>PAYLOAD</diagram></mxfile>
                match = re.search(r"<diagram[^>]*>([\s\S]*?)</diagram>", mxfile_xml)
                if match:
                    return _decode_diagram_payload(match.group(1))
                # Fallback: return the raw mxfile XML
                return mxfile_xml

        pos += 12 + length

    return None


def _strip_ztxt_chunks(data: bytes, key: bytes) -> bytes:
    """Remove all zTXt chunks with the given key from PNG data."""
    result = data[:8]  # PNG signature
    pos = 8
    while pos + 8 <= len(data):
        length = struct.unpack(">I", data[pos : pos + 4])[0]
        chunk_type = data[pos + 4 : pos + 8]
        chunk_end = pos + 12 + length

        keep = True
        if chunk_type == b"zTXt":
            chunk_data = data[pos + 8 : pos + 8 + length]
            null_pos = chunk_data.find(b"\x00")
            if null_pos >= 0 and chunk_data[:null_pos] == key:
                keep = False

        if keep:
            result += data[pos:chunk_end]
        pos = chunk_end

    return result


def _read_bytes(path: str) -> bytes:
    with open(path, "rb") as f:
        return f.read()


# ============================================================
# Export helper
# ============================================================


def export_drawio(drawio_path: str, png_path: str) -> None:
    """Export a .drawio file to PNG using npx draw.io-export."""
    result = subprocess.run(
        ["npx", "draw.io-export", drawio_path, "-o", png_path],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(
            f"npx draw.io-export failed (exit {result.returncode}): {result.stderr}"
        )


# ============================================================
# Subcommands
# ============================================================


def cmd_create(args: argparse.Namespace) -> int:
    xml = _get_xml(args)
    output_path: str = args.output

    if not output_path.endswith(".drawio.png"):
        output_path += ".drawio.png"

    dir_name = os.path.dirname(output_path) or "."
    base_name = os.path.basename(output_path).removesuffix(".drawio.png").removesuffix(".png")
    drawio_path = os.path.join(dir_name, f"{base_name}.drawio")

    # 1. Write .drawio file
    with open(drawio_path, "w", encoding="utf-8") as f:
        f.write(xml)
    print(f"Wrote {drawio_path}", file=sys.stderr)

    # 2. Export to PNG
    try:
        export_drawio(drawio_path, output_path)
        print(f"Exported PNG: {output_path}", file=sys.stderr)
    except RuntimeError as e:
        print(str(e), file=sys.stderr)
        print(json.dumps({"error": str(e), "drawio_path": drawio_path}))
        return 1

    # 3. Embed XML into PNG
    embed_xml_in_png(output_path, xml)
    print(f"Embedded XML into PNG", file=sys.stderr)

    # 4. Clean up .drawio file
    try:
        os.unlink(drawio_path)
    except OSError:
        pass

    print(json.dumps({"path": output_path, "status": "created"}))
    return 0


def cmd_read(args: argparse.Namespace) -> int:
    png_path: str = args.png_path

    if not os.path.exists(png_path):
        print(json.dumps({"error": f"File not found: {png_path}"}))
        return 1

    xml = extract_xml_from_png(png_path)
    if xml is None:
        print(json.dumps({"error": f"No embedded draw.io XML found in: {png_path}"}))
        return 1

    # Output raw XML to stdout (not JSON-wrapped -- it's XML)
    print(xml)
    return 0


def cmd_update(args: argparse.Namespace) -> int:
    xml = _get_xml(args)
    png_path: str = args.png_path

    if not os.path.exists(png_path):
        print(json.dumps({"error": f"File not found: {png_path}"}))
        return 1

    dir_name = os.path.dirname(png_path) or "."
    base_name = os.path.basename(png_path).removesuffix(".drawio.png").removesuffix(".png")
    drawio_path = os.path.join(dir_name, f"{base_name}.tmp.drawio")

    # 1. Write updated .drawio
    with open(drawio_path, "w", encoding="utf-8") as f:
        f.write(xml)

    # 2. Re-export PNG
    try:
        export_drawio(drawio_path, png_path)
        print(f"Re-exported PNG: {png_path}", file=sys.stderr)
    except RuntimeError as e:
        _cleanup(drawio_path)
        print(json.dumps({"error": str(e)}))
        return 1

    # 3. Re-embed XML
    embed_xml_in_png(png_path, xml)
    print(f"Re-embedded XML into PNG", file=sys.stderr)

    # 4. Clean up temp
    _cleanup(drawio_path)

    print(json.dumps({"path": png_path, "status": "updated"}))
    return 0


def _get_xml(args: argparse.Namespace) -> str:
    """Get XML from --xml or --xml-file argument."""
    if hasattr(args, "xml_file") and args.xml_file:
        with open(args.xml_file, "r", encoding="utf-8") as f:
            return f.read()
    if hasattr(args, "xml") and args.xml:
        return args.xml
    # Read from stdin if neither provided
    if not sys.stdin.isatty():
        return sys.stdin.read()
    print("Error: provide XML via --xml, --xml-file, or stdin", file=sys.stderr)
    sys.exit(1)


def _cleanup(path: str) -> None:
    try:
        os.unlink(path)
    except OSError:
        pass


# ============================================================
# CLI
# ============================================================


def main() -> int:
    parser = argparse.ArgumentParser(
        description="draw.io diagram tool -- create, read, and update .drawio.png files",
        epilog="Requires `npx draw.io-export` for PNG rendering.",
    )
    sub = parser.add_subparsers(dest="command", required=True)

    # create
    p_create = sub.add_parser("create", help="Create a new .drawio.png from XML")
    p_create.add_argument("--xml", help="mxGraphModel XML string")
    p_create.add_argument("--xml-file", help="Path to file containing mxGraphModel XML")
    p_create.add_argument("-o", "--output", required=True, help="Output path (should end in .drawio.png)")

    # read
    p_read = sub.add_parser("read", help="Extract draw.io XML from a .drawio.png")
    p_read.add_argument("png_path", help="Path to .drawio.png file")

    # update
    p_update = sub.add_parser("update", help="Update an existing .drawio.png with new XML")
    p_update.add_argument("--xml", help="Updated mxGraphModel XML string")
    p_update.add_argument("--xml-file", help="Path to file containing updated XML")
    p_update.add_argument("png_path", help="Path to existing .drawio.png file")

    args = parser.parse_args()

    if args.command == "create":
        return cmd_create(args)
    elif args.command == "read":
        return cmd_read(args)
    elif args.command == "update":
        return cmd_update(args)
    else:
        parser.print_help()
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: .opencode/tools/drawio.ts
import { tool } from "@opencode-ai/plugin"
import { readFileSync, writeFileSync, unlinkSync, existsSync } from "fs"
import { join, basename, dirname } from "path"
import { deflateRawSync, inflateRawSync, deflateSync, inflateSync } from "zlib"

// ============================================================
// PNG chunk helpers for embedding/extracting draw.io XML
//
// draw.io PNG format (from pzl/drawio-read and draw.io source):
//   zTXt chunk key: "mxGraphModel"
//   zTXt value (after zlib inflate): <mxfile><diagram>PAYLOAD</diagram></mxfile>
//   PAYLOAD: base64( deflateRaw( encodeURIComponent(mxGraphModelXml) ) )
//
// deflateRaw = raw deflate with NO zlib header (wbits=-15)
// ============================================================

function crc32(buf: Buffer): number {
  let crc = 0xffffffff
  for (let i = 0; i < buf.length; i++) {
    crc ^= buf[i]
    for (let j = 0; j < 8; j++) {
      crc = (crc >>> 1) ^ (crc & 1 ? 0xedb88320 : 0)
    }
  }
  return (crc ^ 0xffffffff) >>> 0
}

function encodeDiagramPayload(xml: string): string {
  // mxGraphModel XML → URL-encode → raw deflate → base64
  const urlEncoded = encodeURIComponent(xml)
  const deflated = deflateRawSync(Buffer.from(urlEncoded, "utf-8"))
  return deflated.toString("base64")
}

function decodeDiagramPayload(payload: string): string {
  // base64 → raw inflate → URL-decode → mxGraphModel XML
  const deflated = Buffer.from(payload, "base64")
  const urlEncoded = inflateRawSync(deflated).toString("utf-8")
  return decodeURIComponent(urlEncoded)
}

function embedXmlInPng(pngPath: string, xml: string): void {
  const data = readFileSync(pngPath)

  // Build the mxfile wrapper with encoded diagram payload
  const payload = encodeDiagramPayload(xml)
  const mxfileXml = `<mxfile><diagram>${payload}</diagram></mxfile>`

  // Build zTXt chunk: key="mxGraphModel\0", compression=0, zlib-compressed mxfile XML
  const key = Buffer.from("mxGraphModel\0", "ascii")
  const compressionMethod = Buffer.from([0]) // zlib
  const compressed = deflateSync(Buffer.from(mxfileXml, "utf-8"))
  const chunkData = Buffer.concat([key, compressionMethod, compressed])

  const chunkType = Buffer.from("zTXt", "ascii")
  const lengthBuf = Buffer.alloc(4)
  lengthBuf.writeUInt32BE(chunkData.length, 0)

  const crcInput = Buffer.concat([chunkType, chunkData])
  const crcBuf = Buffer.alloc(4)
  crcBuf.writeUInt32BE(crc32(crcInput), 0)

  const chunk = Buffer.concat([lengthBuf, chunkType, chunkData, crcBuf])

  // Insert the zTXt chunk before IEND (last 12 bytes of a valid PNG)
  const iendStart = data.length - 12
  const before = data.subarray(0, iendStart)
  const iend = data.subarray(iendStart)

  writeFileSync(pngPath, Buffer.concat([before, chunk, iend]))
}

function extractXmlFromPng(pngPath: string): string | null {
  const data = readFileSync(pngPath)

  // Walk PNG chunks looking for zTXt with key "mxGraphModel"
  let pos = 8 // skip PNG signature
  while (pos < data.length) {
    const length = data.readUInt32BE(pos)
    const chunkType = data.subarray(pos + 4, pos + 8).toString("ascii")
    const chunkData = data.subarray(pos + 8, pos + 8 + length)

    if (chunkType === "zTXt") {
      const nullPos = chunkData.indexOf(0)
      const key = chunkData.subarray(0, nullPos).toString("ascii")
      if (key === "mxGraphModel") {
        // byte after null is compression method (0), then zlib-compressed data
        const compressed = chunkData.subarray(nullPos + 2)
        const mxfileXml = inflateSync(compressed).toString("utf-8")
        // Parse the <mxfile><diagram>PAYLOAD</diagram></mxfile> wrapper
        const match = mxfileXml.match(/<diagram[^>]*>([\s\S]*?)<\/diagram>/)
        if (match) {
          return decodeDiagramPayload(match[1])
        }
        // Fallback: return the raw mxfile XML if no diagram tag found
        return mxfileXml
      }
    }

    pos += 12 + length
  }
  return null
}

// ============================================================
// Export helper using npx draw.io-export
// ============================================================

async function exportDrawio(drawioPath: string, pngPath: string): Promise<string | null> {
  try {
    const proc = Bun.spawn(["npx", "draw.io-export", drawioPath, "-o", pngPath], {
      stdout: "pipe",
      stderr: "pipe",
    })
    const exitCode = await proc.exited
    if (exitCode !== 0) {
      const stderr = await new Response(proc.stderr).text()
      return `Export failed (exit ${exitCode}): ${stderr}`
    }
    return null // success
  } catch (err) {
    return `Export failed: ${err instanceof Error ? err.message : String(err)}`
  }
}

// ============================================================
// Tools
// ============================================================

export const create_diagram = tool({
  description: `Create a draw.io diagram from XML and export it as a PNG with the diagram XML embedded inside the image (editable by opening the PNG in draw.io).

Workflow:
1. Writes the XML to a temporary .drawio file
2. Exports to PNG via npx draw.io-export
3. Embeds the source XML into the PNG as a zTXt chunk in draw.io's native format (key: "mxGraphModel")
4. Cleans up the intermediate .drawio file
5. Returns the path to the .drawio.png file

The resulting PNG is viewable as a normal image AND editable in draw.io (drag the PNG onto draw.io to recover the full diagram).`,
  args: {
    xml: tool.schema.string().describe("The draw.io mxGraphModel XML content"),
    output_path: tool.schema.string().describe("Output file path (should end in .drawio.png)"),
  },
  async execute(args) {
    const { xml, output_path } = args

    const dir = dirname(output_path)
    const name = basename(output_path, ".drawio.png").replace(/\.png$/, "")
    const drawioPath = join(dir, `${name}.drawio`)
    const pngPath = output_path.endsWith(".drawio.png") ? output_path : `${output_path}.drawio.png`

    // 1. Write .drawio file
    writeFileSync(drawioPath, xml, "utf-8")

    // 2. Export to PNG
    const exportErr = await exportDrawio(drawioPath, pngPath)
    if (exportErr) {
      return `Error: ${exportErr}\n\nThe .drawio file was saved at: ${drawioPath}\nYou can open it in draw.io manually.`
    }

    // 3. Embed XML into PNG
    try {
      embedXmlInPng(pngPath, xml)
    } catch (err) {
      return `PNG exported but XML embedding failed: ${err instanceof Error ? err.message : String(err)}\n\nPNG saved at: ${pngPath}\n.drawio source at: ${drawioPath}`
    }

    // 4. Clean up .drawio file
    try {
      unlinkSync(drawioPath)
    } catch {}

    return `Diagram created: ${pngPath}\n\nThis PNG has the draw.io XML embedded. Open it in draw.io to edit the diagram.`
  },
})

export const read_diagram = tool({
  description: `Extract the draw.io XML source from an existing .drawio.png file.

Reads the embedded mxfile XML from the PNG's zTXt metadata chunk. Returns the XML so it can be modified and written back with update_diagram.`,
  args: {
    png_path: tool.schema.string().describe("Path to the .drawio.png file"),
  },
  async execute(args) {
    const { png_path } = args

    if (!existsSync(png_path)) {
      return `File not found: ${png_path}`
    }

    const xml = extractXmlFromPng(png_path)
    if (!xml) {
      return `No embedded draw.io XML found in: ${png_path}\n\nThis PNG may not have been created with embedded diagram data.`
    }

    return xml
  },
})

export const update_diagram = tool({
  description: `Update an existing .drawio.png file with new XML. 

Overwrites both the PNG image and the embedded XML source. Use read_diagram first to get the current XML, modify it, then call this to save the changes.`,
  args: {
    xml: tool.schema.string().describe("The updated draw.io mxGraphModel XML content"),
    png_path: tool.schema.string().describe("Path to the existing .drawio.png file to update"),
  },
  async execute(args) {
    const { xml, png_path } = args

    const dir = dirname(png_path)
    const name = basename(png_path, ".drawio.png").replace(/\.png$/, "")
    const drawioPath = join(dir, `${name}.tmp.drawio`)

    // 1. Write updated .drawio
    writeFileSync(drawioPath, xml, "utf-8")

    // 2. Re-export PNG
    const exportErr = await exportDrawio(drawioPath, png_path)
    if (exportErr) {
      try { unlinkSync(drawioPath) } catch {}
      return `Error re-exporting: ${exportErr}`
    }

    // 3. Re-embed XML
    try {
      embedXmlInPng(png_path, xml)
    } catch (err) {
      try { unlinkSync(drawioPath) } catch {}
      return `PNG re-exported but XML embedding failed: ${err instanceof Error ? err.message : String(err)}`
    }

    // 4. Clean up temp
    try { unlinkSync(drawioPath) } catch {}

    return `Diagram updated: ${png_path}`
  },
})
